graylog2-server/src/main/java/org/graylog2/security/realm/CookieAuthenticationRealm.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.security.realm;
+
+import org.apache.shiro.authc.AuthenticationException;
+import org.apache.shiro.authc.AuthenticationInfo;
+import org.apache.shiro.authc.AuthenticationToken;
+import org.apache.shiro.authc.SimpleAccount;
+import org.apache.shiro.authc.credential.AllowAllCredentialsMatcher;
+import org.apache.shiro.realm.AuthenticatingRealm;
+import org.apache.shiro.session.Session;
+import org.apache.shiro.subject.Subject;
+import org.graylog2.plugin.database.users.User;
+import org.graylog2.shared.security.AuthenticationCookieToken;
+import org.graylog2.shared.users.UserService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+
+public class CookieAuthenticationRealm extends AuthenticatingRealm {
```

Comment:
Comparing this with `SessionAuthenticator`, I think we need to be careful to not overlook something.

1. We are not calling `ThreadContext.bind(subject)`. This means, that after the authentication filter logs in the subject [here](https://github.com/Graylog2/graylog2-server/blob/f35df42e165ac570b8b27de3f8eeac85e74ed610/graylog2-server/src/main/java/org/graylog2/shared/security/ShiroAuthenticationFilter.java#L48), the session is not available via `Subject#getSession` anymore. We should make extra sure that this is OK before leaving it out.
2. Session extension was left out. Was this intentional?
3. SSO via trusted headers is not working anymore.

I'm wondering if we should maybe go the safe route and just use the `SessionAuthenticator` and hand it a `SessionIdToken` that we create by extracting the session ID from the cookie.

Reply:
> I'm wondering if we should maybe go the safe route and just use the SessionAuthenticator and hand it a SessionIdToken that we create by extracting the session ID from the cookie. 

Yes, I came to that conclusion after having a second look. I simplified the code to just extract the cookie's value (if existing) before creating the token. This way we can reuse `SessionAuthenticator`, avoid duplication and make sure that the relevant cases are handled.